rabbitmq/retryable_producer.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2019 SumUp Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rabbitmq
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/palantir/stacktrace"
+	"github.com/streadway/amqp"
+	"go.uber.org/zap"
+
+	"github.com/sumup-oss/go-pkgs/backoff"
+	"github.com/sumup-oss/go-pkgs/logger"
+)
+
+type RetryableProducer struct {
+	config        RetryableProducerConfig
+	logger        logger.StructuredLogger
+	metric        Metric
+	clientFactory func(ctx context.Context, config *ClientConfig) (RabbitMQClientInterface, error)
+	cancel        context.CancelFunc
+
+	// mu protects the producer property
+	mu       sync.RWMutex
+	producer *Producer
+}
+
+type RetryableProducerConfig struct {
+	MaxRetryAttempts int
+	// HealthCheckFactor is a number representing how much N multiplied by backoffConfig.Max time is needed
+	// for a block of code to run w/o returning an error, to consider it healthy.
+	// E.g backConfig.Max = 1min, healthCheckFactor = 2, means that code needs to run 2min at least to be healthy
+	// and retried again starting from backoffConfig.Base the next time it has an error.
+	HealthCheckFactor  int
+	BackoffConfig      *backoff.Config
+	RabbitClientConfig *ClientConfig
+}
+
+func NewRetryableProducer(
+	newClientFactory func(ctx context.Context, config *ClientConfig) (RabbitMQClientInterface, error),
+	config RetryableProducerConfig,
+	logger logger.StructuredLogger,
+	metric Metric,
+) *RetryableProducer {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	retryableProducer := &RetryableProducer{
+		logger:        logger,
+		metric:        metric,
+		clientFactory: newClientFactory,
+		config:        config,
+		cancel:        cancel,
+	}
+
+	go retryableProducer.initProducer(ctx)
+
+	return retryableProducer
+}
+
+func (p *RetryableProducer) Publish(
+	exchange,
+	key string,
+	mandatory,
+	immediate bool,
+	expiration string,
+	body []byte,
+	args amqp.Table,
+) error {
+	p.mu.RLock()
+	producer := p.producer
+	p.mu.RUnlock()
```

Comment:
I may be missing something here, but shouldn't we have ` defer p.mu.RUnlock()` here instead? Isn't there a chance for the producer to get swapped or set to nil just after the nil check you have below and before you call `producer.Publish`?

Reply:
You work with the local copy of the producer then. Even if swapped you have it locally, not affected IMO.